src/mlpack/methods/ann/vanilla_rbm.hpp

```diff
@@ -0,0 +1,224 @@
+
+#include "activation_functions/softplus_function.hpp"
+#include "init_rules/gaussian_init.hpp"
+#include "init_rules/random_init.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InitializationRuleType,
+    typename VisibleLayerType,
+    typename HiddenLayerType>
+class RBM
+{
+ public:
+  using NetworkType = RBM<InitializationRuleType, VisibleLayerType,
+      HiddenLayerType>;
+
+  /* 
+   * Intalise all the parameters of the network
+   * using the intialise rule. 
+   *
+   * @tparam: IntialiserType rule to intialise the parameters of the network
+   * @tparam: VisibleLayerType visible layer 
+   * @tparam: HiddenLyaerType hidden layer
+   */
+  RBM(arma::mat predictors, InitializationRuleType initializeRule,
+      VisibleLayerType visible,
+      HiddenLayerType hidden,
+      const size_t numSteps = 1,
+      const bool persistence = false);
+
+  // Reset the network
+  void Reset();
+
+  /* 
+   * Train the netwrok using the Opitimzer with given set of args.
+   * the optimiser sets the paratmeters of the network for providing
+   * most likely parameters given the inputs
+   * @param: predictors data points
+   * @param: optimizer Optimizer type
+   * @param: Args arguments for the optimizers
+   */
+  template<template <typename> class Optimizer>
+  void Train(const arma::mat& predictors, Optimizer<NetworkType>& optimizer);
+
+ /**
+  * Evaluate the rbm network with the given parameters. This function
+  * is usually called by the optimizer to train the model.
+  * The function computes the pseudo likelihood
+  *
+  * @param parameters Matrix model parameters.
+  * @param i Index of point to use for objective function evaluation.
+  * 
+  */
+  double Evaluate(const arma::mat& parameters, const size_t i);
+
+  /**
+  * Monitor Cost this function is needed for checking
+  * the progress of the training. Cross-Entropy is 
+  * needed when peristence is false and pseudo-likelihood
+  * is needed when persistence is true
+  *
+  * @param i Index of point to use for objective function evaluation.
+  * 
+  */
+  double MonitoringCost(const size_t i);
+
+ /** 
+  * This function calculates
+  * the free energy of the model
+  * @param: input data point 
+  */
+  double FreeEnergy(arma::mat input);
+
+ /*
+  * This functions samples the hidden
+  * layer given the visible layer
+  *
+  * @param input visible layer input
+  * @param output the sampled hidden layer
+  */
+  void SampleHidden(arma::mat&& input, arma::mat&& output);
+
+  /*
+  * This functions samples the visble
+  * layer given the hidden layer
+  *
+  * @param input hidden layer
+  * @param output the sampled visible layer 
+  */
+  void SampleVisible(arma::mat&& input, arma::mat&& output);
+
+ /*
+  * This function does the k-step
+  * gibbs sampling.
+  *
+  * @param input: input to the gibbs function
+  * @param output: stores the negative sample
+  */
+  void Gibbs(arma::mat input, arma::mat&& output, size_t steps = 1);
```

Comment:
Here we need this since input is changed if you lookt at the implementation of gibbs.